### DISCUSSION TITLE
ZMQ + GNU Radio send to a hardware

### DISCUSSION THREAD

**Discussion by undeathkhaoz** [srsran/srsRAN_Project#174] [General Help]
Hi! i'm trying to use srsRAN send data to GNU Radio to create some data processing before use an USRP, through ZMQ Rep and Sink, like an image below:

![ZMQwPhysicalRadio](https://github.com/srsran/srsRAN_Project/assets/138717660/663ea4ea-7497-4d5b-86ec-03535f7263d2)

Is this possible using ZMQ?

**Comment by xavierarteaga** [comment]
Hi!

The UE or eNb needs a strict timing constraint of 4 milliseconds between receiving and transmitting.

The ZMQ module achieves this by transmitting 4ms of zeros in the beginning. This makes sure there are always 4ms of buffered samples between the transmitter and the receiver. UHD uses timestamps for indicating the transmit time in the USRP. The safest way to operate USRPs is first to receive a block of data and then transmit another block of data based on the timestamp of the received block. This makes it possible to control the radio latency even if there are instabilities (overflow, underflow, etc).

To sum up, it could be possible if you could control deterministically the UHD latency in GNU radio.

In addition to the deterministic latency issue, you must consider that ZMQ uses internal threads, and dynamic memory which can make extremely difficult the real-time operation.

If I was you, I would:
1. implement the algorithm in C++ (or interconnect GNU Radio blocks in C++);
2. optionally, make a GNU radio block for your algorithm validation; 
3. import your GNU Radio in srsRAN and adapt the interface to the class in `radio.h` with a `radio_wrapper.h`.

**Reply by undeathkhaoz** [reply]
Hello Xavier,

I'm trying to apply a C++ code in the srsRAN_Project, which will be applied before sending the data blocks/buffers to the USRP.

I think that the specific block would vary according to the system's sampling rate, such as 23.04 and 30.72 MS/s, both for TX and RX, but in this case, I would use the config.srate data for dynamic allocation of the C++ module.

With that, I would like to know, if you have this knowledge, where can I obtain these data blocks to be transmitted, which are generated between PHY-LOW and the USRP B210 to apply in a C++ module?

**Comment by pgawlowicz** [comment]
Hi @undeathkhaoz, 
You can have a look at how the channel emulator is implemented our srsRAN_4G project -  https://github.com/srsran/srsRAN_4G/blob/master/lib/src/phy/channel/channel.cc

In srsENB, the DL channel emulator is attached just before samples are sent to the radio device for TX, while in UL before the signal samples enter the PHY layer for demodulation and decoding.
